### DISCUSSION TITLE
notcurses-info and XTSMGRAPHICS

### DISCUSSION THREAD

**Discussion by christianparpart** [dankamongmen/notcurses#2065] [Q&A]
Hey @dankamongmen san,

while making sure your awesome project works better on every commit of contour I realized that `notcurses-info` is setting the sixel color reg count to 256 and then querying it, and then printing the queried result.

So in case of contour this results in having printed 256, whereas on other TEs I realized you see awesome large numbers like 1024 (mlterm) , or even 65536 (wezterm).

I wonder why you query'n'print that value after you have set this yourself. While I might have a wild imagination on the why, wouldn't it be better to print the current max and limit values and then whether or not changing the max-value does have any effect instead?

Just wondering, cheers,
Christian Parpart.

**Comment by dankamongmen** [comment]
Christian Parpart left as an exercise for the reader:
> while making sure your awesome project works better on every
> commit of contour I realized that `notcurses-info` is setting
> the sixel color reg count to 256 and then querying it, and
> then printing the queried result.

indeed!

> So in case of contour this results in having printed 256,
> whereas on other TEs I realized you see awesome large numbers
> like 1024 (mlterm) , or even 65536 (wezterm).

they are churlish, and these values become less awesome the
larger they grow. read on.

> I wonder why you query'n'print that value after you have set
> this yourself. While I might have a wild imagination on the
> why, wouldn't it be better to print the current max and limit
> values and then whether or not changing the max-value does
> have any effect instead?

my quantization algorithm for sixel requires 64 color registers,
and recommends 256. these are both properties of my algorithm
and very likely to change, to whatever degree i spend effort on
improving sixel (i regard sixel as fundamentally inferior to
kitty, but i still want good sixel output. my current encoder is
sub-libpixel on many images, though it does work for all images
with which i've tested it, unlike libsixel).

it does not make use of more than 256, as both palette bandwidth
and my algorithm's runtime are O(N) on the number of quantized
colors. i only devote a single byte to the color in internal
Sixel-related data structures, though that's obviously easy
enough to change. supporting more than 256 for a Notcurses
application is thus a waste of space, assuming the terminal
requires space O(lgN) on the palette size (if it lacks such a
dependency, there's no waste, no mess, no fuss).

i need to know the number supported, so i don't try to emit 256
colors to a 64-count palette. the terminal starts supporting N
colors, where N <= P, the possible number of colors.

if N < 256 but P > N, i want to raise N to max(P, 256).

accomplishing this augmentation of N and knowing the final value
of N is most cheaply accomplished in 2 directives. proof is left
to the reader.

sending a signal to the terminal emulator that i'll never need
more than 256 color registers is a side effect, which the
emulator is free to make use of if it desires.

WezTerm claims 65,536 registers. gcc supports -O9. Kempai, do
not allow yourself, nor your code, to be judged by such empty
numbers. 256 color registers is, for now, the appropriate number.

rigorously, nick

-- 
nick black -=- https://www.nick-black.com
to make an apple pie from scratch,
you need first invent a universe.

**Reply by dankamongmen** [reply]
but by all means don't allow this to keep you from settling the question with @wez in the streets

**Reply by dankamongmen** [reply]
> accomplishing this augmentation of N and knowing the final value of N is most cheaply accomplished in 2 directives. proof is left to the reader.

remember that this happens during initialization, and thus is a cost in every single notcurses context ever. when people are benchmarking `ncplayer` run one-argument-per-process like a bunch of one-eyed cave-dwelling batpeople, any costs here fuck me right up the ass. i'm absolutely resolved to keeping initialization down to a single RTT between myself and the terminal. no lengthy TLS-like dialogues asking one another where we grew up and what endianness our dicks are and all that

**Reply by wez** [reply]
FWIW, WezTerm's sixel color map is simplistically implemented as a `HashMap<u16, RgbColor>`.
I haven't done anything in particular to optimize the sixel decoding in wezterm so there may be a bit of room for improvement. I don't expect that part to be the bottleneck in my implementation TBH!

**Reply by dankamongmen** [reply]
i don't think i stated this as clearly as i'd have liked to. also, i erroneously stated that i require 64 cregs; i can work with as few as 16.

so there are three cases going in: the terminal has N=256, less than 256, or more than 256. in addition, it might support some larger number P, in which case it presumably supports setting N as high as P. it might further support reducing N to some floor M.

all i care about is that i get 256 if 256 are available. i thus send a SET 256 *unconditionally*, not querying for N ahead of time. if N is greater than or equal to 256, great! it'll remain greater than or equal to 256, though i won't be taking advantage of the color registers >= 256. if N was less than 256, it is now 256 if that is possible. immediately following this directive (i.e. without waiting for a response) i send a QUERY N. i only care about the final value, not the original one.

i'm guessing that contour honors the reductive SET, whereas these others do not. to me, that just makes contour look lean and mean. if you're worried it casts contour in an unfair light, i can change `notcurses-info` to print the number of cregs up through the number we intend to use (i.e. cap it at 256). but that's something of a fundamentally untrue flattening of knowledge, and if someone ever like shelled out from notcurses and generated a sixel directly to the terminal, they would actually have 64K to work with in one and 256 in the other, so yeah nevermind i don't want to change that. if it bugs you, don't honor my reductive SET. i'm not a cop.

the end result is that i have 256 registers if at all possible, and i know how many are available, in a single RTT.

in the words of Marlo Stanfield, "i just want 'em moved"

[![i just want em moved](http://img.youtube.com/vi/nvy3QPZlZcc/0.jpg)](http://www.youtube.com/watch?v=nvy3QPZlZcc "i just want em moved")
